flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyV2.java

```diff
@@ -28,16 +51,210 @@
+	 * Derived classes can override this method to customize the client configuration.
+	 *
+	 * @param configProps the properties map used to create the Kinesis Client
+	 * @return a Kinesis Client
+	 */
+	protected KinesisAsyncClient createKinesisAsyncClient(final Properties configProps) {
+		final ClientConfiguration config = new ClientConfigurationFactory().getConfig();
+		return AwsV2Util.createKinesisAsyncClient(configProps, config);
+	}
+
+	/**
+	 * {@inheritDoc}
 	 */
-	public KinesisProxyV2(final KinesisAsyncClient kinesisAsyncClient) {
-		this.kinesisAsyncClient = kinesisAsyncClient;
+	@Override
+	public Map<String, String> describeStream(List<String> streams) throws InterruptedException, ExecutionException {
+		Map<String, String> result = new HashMap<>();
+		for (String stream : streams) {
+			DescribeStreamRequest describeStreamRequest = DescribeStreamRequest
+				.builder()
+				.streamName(stream)
+				.build();
+			DescribeStreamResponse describeStreamResponse = null;
+
+			int retryCount = 0;
+			while (retryCount <= fanOutRecordPublisherConfiguration.getDescribeStreamMaxRetries() && describeStreamResponse == null) {
+				try {
+					describeStreamResponse = kinesisAsyncClient.describeStream(describeStreamRequest).get();
+				} catch (ExecutionException ex) {
+					if (AwsV2Util.isRecoverableException(ex)) {
+						long backoffMillis = fullJitterBackoff(
+							fanOutRecordPublisherConfiguration.getDescribeStreamBaseBackoffMillis(), fanOutRecordPublisherConfiguration.getDescribeStreamMaxBackoffMillis(), fanOutRecordPublisherConfiguration.getDescribeStreamExpConstant(), retryCount++);
+						LOG.warn("Got recoverable AmazonServiceException when trying to describe stream " + stream + ". Backing off for "
+							+ backoffMillis + " millis (" + ex.getClass().getName() + ": " + ex.getMessage() + ")");
+						Thread.sleep(backoffMillis);
+					} else {
+						throw ex;
+					}
+				}
+			}
+			if (describeStreamResponse == null) {
+				throw new RuntimeException("Retries exceeded for describeStream operation - all " + fanOutRecordPublisherConfiguration.getDescribeStreamMaxRetries() +
+					" retry attempts failed.");
+			}
+			result.put(stream, describeStreamResponse.streamDescription().streamARN());
+		}
+		return result;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public List<FanOutStreamInfo> registerStreamConsumer(Map<String, String> streamArns) throws InterruptedException, ExecutionException {
+		Preconditions.checkArgument(fanOutRecordPublisherConfiguration.getConsumerName().isPresent());
+		String consumerName = fanOutRecordPublisherConfiguration.getConsumerName().get();
+		List<FanOutStreamInfo> result = new ArrayList<>();
+		for (Map.Entry<String, String> entry : streamArns.entrySet()) {
+			String stream = entry.getKey();
+			String streamArn = entry.getValue();
+			RegisterStreamConsumerRequest registerStreamConsumerRequest = RegisterStreamConsumerRequest
```

Comment:
@xiaolong-sn what happens if the stream consumer is already registered? The flow chart in the FLIP had a call to `ListStreamConsumers` to establish if it is already registered. Also in the case of Lazy we are meant to be introducing a jitter delay up front. Please refer to the flowcharts in the [FLIP](https://cwiki.apache.org/confluence/display/FLINK/FLIP-128%3A+Enhanced+Fan+Out+for+AWS+Kinesis+Consumers)

Reply:
Ah I see you are using the `describeStreamConsumer` instead of list. That is better actually. But I think you should do that first, as per the flow chart. Just replace List with Describe. If the consumer is already registered it will be 1 call instead of 2 per task. 

Describe has a higher TPS too so that is good!

Reply:
I changed the interface of `registerStreamConsumer`, it's my fault that I did not fully understand the FLIP before.